File: go-src/internal/storage/storage.go
// Package storage manages VecFS JSONL file storage: store, search, updateScore, delete.
// Matches ts-src/storage.ts behaviour.
package storage

import (
	"bufio"
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/WazzaMo/vecfs/internal/sparse"
)

const feedbackRankWeight = 0.1

// VecFSEntry is a single stored entry (matches TS VecFSEntry).
type VecFSEntry struct {
	ID        string            `json:"id"`
	Metadata  map[string]any    `json:"metadata"`
	Vector    sparse.Vector     `json:"vector"`
	Score     float64           `json:"score"`
	Timestamp int64             `json:"timestamp"`
}

// SearchResult extends an entry with similarity (matches TS SearchResult).
type SearchResult struct {
	VecFSEntry
	Similarity float64 `json:"similarity"`
}

// Storage manages a JSONL file with mutex-protected access.
type Storage struct {
	filePath    string
	entries     []*VecFSEntry
	initialized  bool
	mu          sync.Mutex
}

// New creates a Storage that uses the given file path.
func New(filePath string) *Storage {
	return &Storage{filePath: filePath}
}

func feedbackBoost(score float64) float64 {
	abs := score
	if abs < 0 {
		abs = -abs
	}
	return feedbackRankWeight * (score / (1 + abs))
}

func combinedRank(similarity, score float64) float64 {
	return similarity + feedbackBoost(score)
}

func (s *Storage) ensureFileUnlocked() error {
	if s.initialized {
		return nil
	}
	dir := filepath.Dir(s.filePath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	if _, err := os.Stat(s.filePath); os.IsNotExist(err) {
		if err := os.WriteFile(s.filePath, nil, 0644); err != nil {
			return err
		}
	}
	s.initialized = true
	return nil
}

// EnsureFile creates the file and parent dir if needed. Safe to call multiple times.
func (s *Storage) EnsureFile() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.ensureFileUnlocked()
}

func (s *Storage) loadEntriesLocked() ([]*VecFSEntry, error) {
	if s.entries != nil {
		return s.entries, nil
	}
	if err := s.ensureFileUnlocked(); err != nil {
		return nil, err
	}
	data, err := os.ReadFile(s.filePath)
	if err != nil {
		return nil, err
	}
	var entries []*VecFSEntry
	scanner := bufio.NewScanner(bufio.NewReader(bytes.NewReader(data)))
	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}
		var e VecFSEntry
		if err := json.Unmarshal(line, &e); err != nil {
			continue // skip malformed
		}
		entries = append(entries, &e)
	}
	s.entries = entries
	return s.entries, nil
}

func (s *Storage) persistAllLocked(entries []*VecFSEntry) error {
	var buf []byte
	for _, e := range entries {
		line, err := json.Marshal(e)
		if err != nil {
			return err
		}
		buf = append(buf, line...)
		buf = append(buf, '\n')
	}
	return os.WriteFile(s.filePath, buf, 0644)
}

func (s *Storage) persistAppendLocked(entry *VecFSEntry) error {
	f, err := os.OpenFile(s.filePath, os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer f.Close()
	line, err := json.Marshal(entry)
	if err != nil {
		return err
	}
	_, err = f.Write(append(line, '\n'))
	return err
}

// Store upserts an entry. Returns true if new, false if updated.
func (s *Storage) Store(entry *VecFSEntry) (bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	entries, err := s.loadEntriesLocked()
	if err != nil {
		return false, err
	}
	full := *entry
	if full.Timestamp == 0 {
		full.Timestamp = nowMs()
	}
	for i, e := range entries {
		if e.ID == entry.ID {
			entries[i] = &full
			return false, s.persistAllLocked(entries)
		}
	}
	entries = append(entries, &full)
	s.entries = entries
	return true, s.persistAppendLocked(&full)
}

// Search returns entries sorted by combined rank (similarity + feedback boost), limited.
func (s *Storage) Search(query sparse.Vector, limit int) ([]*SearchResult, error) {
	s.mu.Lock()
	entries, err := s.loadEntriesLocked()
	s.mu.Unlock()
	if err != nil {
		return nil, err
	}
	if limit <= 0 {
		limit = 5
	}
	queryNorm := sparse.Norm(query)
	results := make([]*SearchResult, 0, len(entries))
	for _, e := range entries {
		sim := sparse.CosineSimilarity(query, e.Vector, queryNorm)
		results = append(results, &SearchResult{VecFSEntry: *e, Similarity: sim})
	}
	// Sort by combined rank descending
	sortSearchResults(results)
	if len(results) > limit {
		results = results[:limit]
	}
	return results, nil
}

func sortSearchResults(r []*SearchResult) {
	// Simple sort by combined rank desc
	for i := 0; i < len(r); i++ {
		for j := i + 1; j < len(r); j++ {
			ci := combinedRank(r[i].Similarity, r[i].Score)
			cj := combinedRank(r[j].Similarity, r[j].Score)
			if cj > ci {
				r[i], r[j] = r[j], r[i]
			}
		}
	}
}

// UpdateScore adjusts the score of an entry by ID. Returns true if found.
func (s *Storage) UpdateScore(id string, delta float64) (bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	entries, err := s.loadEntriesLocked()
	if err != nil {
		return false, err
	}
	for _, e := range entries {
		if e.ID == id {
			e.Score += delta
			return true, s.persistAllLocked(entries)
		}
	}
	return false, nil
}

// Delete removes an entry by ID. Returns true if found.
func (s *Storage) Delete(id string) (bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	entries, err := s.loadEntriesLocked()
	if err != nil {
		return false, err
	}
	for i, e := range entries {
		if e.ID == id {
			entries = append(entries[:i], entries[i+1:]...)
			s.entries = entries
			return true, s.persistAllLocked(entries)
		}
	}
	return false, nil
}

func nowMs() int64 { return time.Now().UnixMilli() }

File: py-src/vecfs_py/storage.py
"""
VecFS storage: JSONL file with in-memory cache and file mutex.

Matches ts-src/storage.ts and go-src storage: same entry shape and ranking.
"""

from __future__ import annotations

import asyncio
import json
from pathlib import Path
from typing import Any

from .sparse import cosine_similarity, norm

FEEDBACK_RANK_WEIGHT = 0.1


def _feedback_boost(score: float) -> float:
    """Bounded contribution of reinforcement score to ranking."""
    return FEEDBACK_RANK_WEIGHT * (score / (1 + abs(score)))


def _combined_rank(similarity: float, score: float) -> float:
    """Higher is better; used to sort search results."""
    return similarity + _feedback_boost(score)


def _entry_to_jsonl_line(entry: dict[str, Any]) -> str:
    """Serialize entry for JSONL; vector keys as strings for JSON."""
    out = dict(entry)
    if "vector" in out and out["vector"]:
        out["vector"] = {str(k): v for k, v in out["vector"].items()}
    return json.dumps(out) + "\n"


def _parse_line(line: str) -> dict[str, Any] | None:
    """Parse one JSONL line; convert vector keys to int for internal use."""
    line = line.strip()
    if not line:
        return None
    try:
        entry = json.loads(line)
    except json.JSONDecodeError:
        return None
    if "vector" in entry and isinstance(entry["vector"], dict):
        entry["vector"] = {int(k): v for k, v in entry["vector"].items()}
    return entry


class VecFSStorage:
    """
    Manages VecFS entries in a local JSONL file.
    Entries cached in memory; mutations persist under a mutex.
    """

    def __init__(self, file_path: str) -> None:
        self._path = Path(file_path)
        self._entries: list[dict[str, Any]] | None = None
        self._initialized = False
        self._lock = asyncio.Lock()

    async def ensure_file(self) -> None:
        """Ensure the storage file and parent directory exist."""
        if self._initialized:
            return
        self._path.parent.mkdir(parents=True, exist_ok=True)
        if not self._path.exists():
            self._path.write_text("", encoding="utf-8")
        self._initialized = True

    async def _load_entries(self) -> list[dict[str, Any]]:
        """Load all entries from file into cache."""
        if self._entries is not None:
            return self._entries
        await self.ensure_file()
        text = self._path.read_text(encoding="utf-8")
        self._entries = []
        for line in text.splitlines():
            if not line:
                continue
            entry = _parse_line(line)
            if entry is not None:
                self._entries.append(entry)
        return self._entries

    async def _persist_all(self) -> None:
        """Rewrite the file from the in-memory cache."""
        if self._entries is None:
            return
        content = "".join(_entry_to_jsonl_line(e) for e in self._entries)
        self._path.write_text(content, encoding="utf-8")

    async def _persist_append(self, entry: dict[str, Any]) -> None:
        """Append one entry to the file."""
        with self._path.open("a", encoding="utf-8") as f:
            f.write(_entry_to_jsonl_line(entry))

    async def store(
        self,
        id: str,
        vector: dict[int, float],
        metadata: dict[str, Any] | None = None,
        score: float = 0.0,
    ) -> bool:
        """Upsert an entry. Returns True if new, False if updated."""
        import time

        async with self._lock:
            entries = await self._load_entries()
            full: dict[str, Any] = {
                "id": id,
                "metadata": metadata or {},
                "vector": vector,
                "score": score,
                "timestamp": int(time.time() * 1000),
            }
            for i, e in enumerate(entries):
                if e.get("id") == id:
                    entries[i] = full
                    await self._persist_all()
                    return False
            entries.append(full)
            await self._persist_append(full)
            return True

    async def search(
        self,
        query_vector: dict[int, float],
        limit: int = 5,
    ) -> list[dict[str, Any]]:
        """Return entries sorted by combined rank (similarity + feedback boost)."""
        entries = await self._load_entries()
        query_norm = norm(query_vector)
        results: list[dict[str, Any]] = []
        for e in entries:
            vec = e.get("vector") or {}
            if isinstance(vec, dict) and vec:
                sim = cosine_similarity(
                    query_vector,
                    vec,
                    v1_norm=query_norm,
                )
            else:
                sim = 0.0
            results.append(
                {
                    **e,
                    "similarity": sim,
                }
            )
        results.sort(
            key=lambda r: _combined_rank(r["similarity"], r.get("score") or 0),
            reverse=True,
        )
        return results[:limit]

    async def update_score(self, id: str, score_adjustment: float) -> bool:
        """Adjust reinforcement score for an entry. Returns True if found."""
        async with self._lock:
            entries = await self._load_entries()
            for e in entries:
                if e.get("id") == id:
                    e["score"] = (e.get("score") or 0) + score_adjustment
                    await self._persist_all()
                    return True
            return False

    async def delete(self, id: str) -> bool:
        """Remove an entry by id. Returns True if found."""
        async with self._lock:
            entries = await self._load_entries()
            for i, e in enumerate(entries):
                if e.get("id") == id:
                    entries.pop(i)
                    await self._persist_all()
                    return True
            return False
